### PR TITLE
SILOptimizer: A new "TempLValueOpt" optimization for copy_addr

### PR DESCRIPTION
**Pull request by eeckstein**
Optimizes copies from a temporary (an "l-value") to a destination.
```
    %temp = alloc_stack $Ty
    instructions_which_store_to %temp
    copy_addr [take] %temp to [initialization] %destination
    dealloc_stack %temp
```
is optimized to
```
    instructions_which_store_to %destination
```
The name tempLValueOpt refers to the TempRValueOpt pass, which performs a related transformation, just with the temporary on the "right" side.
The tempLValueOpt is similar to CopyForwarding::backwardPropagateCopy.
It's more restricted (e.g. the copy-source must be an alloc_stack, the copy-destination must be initialized).
That enables other patterns to be optimized, which backwardPropagateCopy cannot handle.

This PR also contains some other small improvements related to copy_addr and other address instructions. For details see the commit messages.

The effect of this optimizations is that code with address-only types (e.g. generic functions) has significantly less copy instructions, especially initializations of address-only enums (e.g. optionals).
This is a benefit for performance and code size and can also reduce the amount of generic metadata which is generated at runtime.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by atrick** [comment]
On Alias Analysis support for enums. This is something alias analysis should definitely handle. I'm not surprised is wasn't handled, since the code is neglected. But it's definitely worth fixing.

With something so fundamental, I think it's good to have negative tests for any complex cases, particularly in aliasAddressProjection, which I'm not sure I fully understand. The way aliasAddressProjection and aliasInner are mutually recursive makes it hard to reason about. We obviously need to be very careful about introducing false positives here.

For example, we should test that
E -> init_existential/enum
aliases with
E -> init_existential/enum -> struct_element_addr

While we're at it:
E -> init_existential/enum -> struct_element_addr #1
does not alias with
E -> init_existential/enum -> struct_element_addr #2

More intersting:
V -> struct_element_addr -> init_existential/enum
aliases with
V -> struct_element_addr -> init_existential/enum

What about the fact that existential storage can actually alias even though the existential values are from different storage addresses? I think you at least need to come up with a proof that this it's generally safe to treat them as non-aliasing. I don't think it's as simple as "it makes sense for alias analysis to look through them". How do we know it's always correct to "look through them"?

**Comment by atrick** [comment]
On `MemoryBehaviorVisitor::visitCopyAddrInst`. This makes sense to me. I'm surprised we didn't have it already. Again, I would have added a couple of negative tests thought... those are really much more important than the positive test!

**Comment by atrick** [comment]
On SILCombiner::combineCopyAndDestroy...

In principle, I don't like scanning the instruction stream within a SILCombineVisitor. SILCombine should only be for SSA-based peepholes. (I have the same criticism of `SILCombiner::visitInjectEnumAddrInst`).

This particular case might be justifiable in practice, since it bails at the first instruction that `mayWriteMemory`, but only if it's really needed by SILCombine. However, I don't see any reason this should be in SILCombine in the first place. Is it needed to enable some other SILCombines? Why would we add a SILCombine for something that's already handled in another pass by a more general analysis?

`combineCopyAndDestroy` is something that should be done during destroy hoisting.

CopyForwarding::hoistDestroy can easily be run as a separate pass. (The implementation is mostly in `CopyForwarding::forwardCopiesOf`--it's being gated by the copy analysis of just to avoid doing extra work.) Then we can do "DestroyHoisting" without paying the huge cost of doing a global dataflow analysis of all memory sublocations simultaneously--and it will handle more cases. Just by adding an AliasAnalysis query, it should be able to handle any case where either (a) we can recognize all address uses (aka known-use) or (b) alias analysis can recognize no potential aliasing access between the copy and destroy (aka known-instruction-window).

CopyForwarding::hoistDestroy is not limited to a single basic block and not limited to a single destroy. Any case that would be handled by `SILCombiner::combineCopyAndDestroy` would trivially be handled by CopyForwarding::hoistDestroy.

CopyForwarding::hoistDestroy currently assumes a "known-use" model. As with most our analyses, we really need both a "known-use" and "known-instruction-window" mode. The known-use mode is great because it's very efficient, its cost does not scale with the size of the function, and it handles arbitrarily complex code. Of course, it doesn't handle any SSA def-use pattern that we haven't specifically taught it to recognize.

A known-instruction-window mode is good for handling many patterns where we can't analyze the a value's source and can't find all aliases in the known-use mode. It's expensive, because it needs Alias and Escape Analysis, but important at -O. (CopyForwarding was originally supposed to be a -Onone peephole).

So, in short, we can just add an AliasAnalysis query to the existing destroy hoisting and easily separate it from `forwardCopiesOf` if need be.

Note that the AliasAnalysis query needs to bail on anything that can either read, write, or release memory, so it's fairly limited, but could still handle cases where the destroy is close to the copy, but we can't analysis the copy's source address for some reason.

Somehow, your current implementation seems to be ignoring cases like this:
```
copy_addr %source to %destination
use %source
destroy_addr %source
```

**Comment by atrick** [comment]
From what I can tell so far, tempLValueOpt is an attempt to add AliasAnalysis to CopyForwarding. I think it's a very bad idea for SILCombine to depend on Alias/Escape Analysis. SILCombine should be SSA peepholes that can rerun frequently after other transforms without recomputing interprocedural analysis. (We really need to do `SILCombiner::visitInjectEnumAddrInst` in a separate pass too).

Scanning the instructions in SILCombine also doesn't make sense to me. That's quadratic even if SILCombine does nothing, and cubic or worse of SILCombine makes changes--multiplied by the cost of the alias analysis queries. I can't even reason about what happens in the presence of other SILCombines that invalidate alias and escape analysis or adds new instructions that are unrecognized by the previous analysis.

It would be fine to just add AliasAnalysis to CopyForwarding, since we're not using it as the -Onone peephole that it was meant to be.

I haven't looked yet at why CopyForwarding doesn't handle these patterns without your change, but there are two reasonable approaches:

1. If the address use analysis currently can't recognize something, add the SSA def-use patterns to visitAddressUse.

OR

2. When we can't analyze all address uses, don't bailout entirely. Instead, just skip AnalyzeBackwardUse, and query AliasAnalysis. If the loop in backwardPropagateCopy is too hard to read, then it would be easy factor it into two separate routines, one with a loop that checks known-uses, and the other with a loop that checks AliasAnalysis. But aside from that, the function of the pass is basically the same.

**Comment by atrick** [comment]
Ok, TempLValue is actually a different algorithm than backwardPropagateCopy. This is really just handling the case of a true "temporary":

```
alloc_stack temp
init temp
copy temp -> dest
dealloc_stack temp
```

...with no other uses of the temp outside the range of instructions from `alloc` to `copy`.

So this could be a separate TempLValue pass. It's easy to extract destroy hoisting out of CopyForwarding since this new pass also requires it. Should I do that?

I'm still not sure why backwardPropagateCopy wouldn't handle these cases if it used AliasAnalysis, but it may make sense to separate passes that depend on AliasAnalysis from those that don't.

**Comment by eeckstein** [comment]
@atrick Thanks for reviewing!

* alias analysis and membehavior: 100% agreed. I added tests. I even found some deficiencies because of the tests and therefore changed the implementation a bit.

* existentials: AFAIK, existentials only share the same memory if they are copied. But initialization with init_existential_addr can only return the operand address or a new allocated memory. And even if we would also handle open_existential_addr, it would be fine, because copy-on-write existentials are only an "implementation detail" of IRGen and this behavior not "observable" in SIL.

* TempLValueOpt: I thought long about if I should make it a separate pass or put it into SILCombine. My laziness won. I agree, it's better as a separate pass. I changed it.

* copyaddr-destroy peephole: I tried to add this to CopyForwarding, but somehow it didn't work because (I think) CopyForwarding only handles destroys which are in a certain relationship with propagated copies (but I might be wrong). I don't fully understand the logic here. Anyway, this peephole is also done by the DestroyHoisting pass, but it only runs on OSSA. Once we have OSSA in the full pipeline we can run DestroyHoisting later and then then we can remove the peephole from TempLValueOpt.

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by atrick** [comment]
This is mostly for the record:

The problem with CopyForwarding is that it's analyzing copy's destination just so it doesn't need to check AliasAnalysis later.

One problem is that findAddressRootAndUsers only looks through one level of `init_enum/existential`, and some of your cases have two levels, which must cause `isIdentifiedDestValue` to fail. That's actually pretty trivial to fix.

But the more basic problem is that `visitAddressUsers` bails out on all kinds of interesting uses. I'm not sure if they are relevant to your cases, but that entire analysis of the destination's lifetime is irrelevant once you use alias analysis.

**Comment by atrick** [comment]
It would be super helpful to describe this analysis and the transformation in a way that someone can reason about. It's simple enough to specify in a comment:

```
alloc_stack %temp
"first use" %temp // (beginning of %temp lifetime)
// no definition of %dest_root
// no writes to %dest
copy_addr [take] %temp to [initialization] %dest // (end of %temp lifetime)
// no uses of %temp
dealloc_stack %temp

All projections between %dest_root and %dest are hoisted above beginning of %temp lifetime.

All uses of %temp are then directly replaced by %dest.
```

Also, explain why this does not handle reads from %dest:

```
alloc_stack %temp
"first use" %temp // (beginning of %temp lifetime)
use %dest
copy_addr [take] %temp to [initialization] %dest // (end of %temp lifetime)
```

Are you just assuming there needs to be a write to %dest before the initialization? I'm not sure where that's guaranteed. I'm not even sure what a "write" has to do with a "destroy". You could either somehow check that %dest isn't be destroyed, or just check for both reads and writes--then you don't care if the copy_addr is an initialization.

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
You are right, destroy_addr is not modeled as "write". I have to check for reads and writes.

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
Looks like we need to model a take (e.g. load [take], copy_addr [take]) as a write to memory. I added a commit which fixes this. It can cause memory lifetime failures in OSSA.

File: lib/SILOptimizer/Transforms/TempLValueOpt.cpp [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+  // memory writes between them.
+  for (auto iter = std::next(copyInst->getIterator());
+       iter != block->end(); ++iter) {
+    SILInstruction *inst = &*iter;
+    if (inst == destroy) {
+      copyInst->setIsTakeOfSrc(IsTake);
+      destroy->eraseFromParent();
+      return true;
+    }
+    if (inst->mayWriteToMemory())
```
Comment (atrick) [inline lib/SILOptimizer/Transforms/TempLValueOpt.cpp:288]: Here's where destroy hoisting does not check for reads. Maybe you didn't push the latest change yet.

Comment (eeckstein) [reply lib/SILOptimizer/Transforms/TempLValueOpt.cpp]: Yes, I forgot to change this. I pushed a new version with a fix

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Review by atrick — APPROVED** [review]
LGTM except for the two comment blocks in destroy hoisting.

**Comment by eeckstein** [comment]
@swift-ci smoke test
